File: jip/conformance/python/conformance_test.py
#!/usr/bin/env python3
# Copyright 2026 J3nna Technologies, LLC
# SPDX-License-Identifier: Apache-2.0
#
# J3nna Mesh wire-conformance test for Python. Proves a Python implementation reproduces the canonical
# signing bytes byte-for-byte and verifies the reference signatures from ../vectors.json. The framing
# helpers here are the seed of the Python SDK's wire layer.
#
#   python3 conformance_test.py        # exits 0 on pass, non-zero on any mismatch
#   pytest conformance_test.py         # also runs as a pytest case
#
# Requires `cryptography` (ed25519). The whole point: if this passes, a Python peer is wire-compatible with
# a Go peer.

import base64
import binascii
import hashlib
import json
import struct
from pathlib import Path

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

VECTORS = Path(__file__).resolve().parent.parent / "vectors.json"


def _field(buf: bytearray, x: bytes) -> None:
    """4-byte big-endian length prefix, then the raw bytes."""
    buf += struct.pack(">I", len(x))
    buf += x


def _u64(buf: bytearray, v: int) -> None:
    """8-byte big-endian integer."""
    buf += struct.pack(">Q", v & 0xFFFFFFFFFFFFFFFF)


def presence_signing_bytes(inp: dict) -> bytes:
    b = bytearray()
    _field(b, inp["protocol"].encode())
    _field(b, inp["alg"].encode())
    _field(b, inp["id"].encode())
    _field(b, bytes.fromhex(inp["public_key_hex"]))
    _field(b, inp["endpoint"].encode())
    _field(b, inp["mcp_path"].encode())
    caps = sorted(inp["capabilities"])
    b += struct.pack(">I", len(caps))
    for c in caps:
        _field(b, c.encode())
    b += struct.pack(">I", int(inp["protocol_major"]))
    _field(b, inp["grant_id"].encode())
    _u64(b, int(inp["heartbeat_unix"]))
    return bytes(b)


def grant_signing_bytes(inp: dict) -> bytes:
    b = bytearray()
    _field(b, b"J3nna-mesh-grant/1")
    _field(b, inp["alg"].encode())
    _field(b, inp["id"].encode())
    _field(b, inp["subject"].encode())
    _field(b, bytes.fromhex(inp["public_key_hex"]))
    _u64(b, int(inp["tier"]))
    _field(b, b"\x00".join(s.encode() for s in sorted(inp["scopes"])))
    _u64(b, int(inp["issued_at"]))
    _u64(b, int(inp["not_after"]))
    if inp.get("principal"):
        _field(b, b"J3nna-mesh-principal/1")
        _field(b, inp["principal"].encode())
    return bytes(b)


def canonical_args_json(args: dict) -> bytes:
    """Reproduce Go's json.Marshal of a map: keys sorted, no insignificant whitespace, and <, >, & escaped
    as \\u003c \\u003e \\u0026 (Go's default HTML escaping)."""
    s = json.dumps(args, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
    s = s.replace("<", "\\u003c").replace(">", "\\u003e").replace("&", "\\u0026")
    return s.encode()


def callproof_signing_bytes(inp: dict) -> bytes:
    b = bytearray()
    _field(b, b"JIP-call/0.2")
    _field(b, inp["alg"].encode())
    _field(b, inp["node_id"].encode())
    _field(b, inp["tool"].encode())
    _field(b, bytes.fromhex(inp["args_hash_hex"]))
    _u64(b, int(inp["unix_milli"]))
    return bytes(b)


BUILDERS = {
    "presence-record": presence_signing_bytes,
    "grant": grant_signing_bytes,
    "callproof": callproof_signing_bytes,
}


def run() -> None:
    data = json.loads(VECTORS.read_text())
    assert data["protocol"] == "JIP/0.1", data["protocol"]
    for v in data["vectors"]:
        name = v["name"]
        builder = BUILDERS.get(name)
        assert builder is not None, f"no Python builder for vector {name!r}"

        # 1. byte-exact framing
        got = builder(v["input"])
        assert binascii.hexlify(got).decode() == v["signing_bytes_hex"], f"{name}: signing bytes differ"

        # 2. signature verifies against the reference public key
        pub = Ed25519PublicKey.from_public_bytes(bytes.fromhex(v["signer_public_key_hex"]))
        sig = base64.b64decode(v["signature_b64"])
        pub.verify(sig, got)  # raises InvalidSignature on mismatch

        # 3. the one cross-language subtlety: the CallProof args canonicalization
        if name == "callproof":
            cj = canonical_args_json(v["input"]["args"]).decode()
            assert cj == v["input"]["args_canonical_json"], "args canonical JSON differs from Go"
            h = hashlib.sha256(canonical_args_json(v["input"]["args"])).hexdigest()
            assert h == v["input"]["args_hash_hex"], "args hash differs"

        print(f"  ok  {name}")
    print(f"PASS: {len(data['vectors'])} vectors verified (Python wire-compatible with the Go reference)")


def test_conformance() -> None:  # pytest entry point
    run()


if __name__ == "__main__":
    run()

File: jip/conformance/node/conformance_test.mjs
// Copyright 2026 J3nna Technologies, LLC
// SPDX-License-Identifier: Apache-2.0
//
// J3nna Mesh wire-conformance test for Node.js / TypeScript. Proves a Node implementation reproduces the
// canonical signing bytes byte-for-byte and verifies the reference signatures from ../vectors.json, using
// only Node's built-in `crypto` (ed25519 — no dependencies). The framing helpers here are the seed of the
// TS/Node SDK's wire layer.
//
//   node conformance_test.mjs          # exits 0 on pass, non-zero on any mismatch

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import crypto from "node:crypto";

const here = dirname(fileURLToPath(import.meta.url));
const vectors = JSON.parse(readFileSync(join(here, "..", "vectors.json"), "utf8"));

// ── framing primitive ──
function field(parts, x) {
  const buf = Buffer.isBuffer(x) ? x : Buffer.from(x, "utf8");
  const len = Buffer.alloc(4);
  len.writeUInt32BE(buf.length, 0);
  parts.push(len, buf);
}
function u64(parts, v) {
  const b = Buffer.alloc(8);
  b.writeBigUInt64BE(BigInt(v), 0);
  parts.push(b);
}
function u32(parts, v) {
  const b = Buffer.alloc(4);
  b.writeUInt32BE(v >>> 0, 0);
  parts.push(b);
}
const hexBuf = (h) => Buffer.from(h, "hex");

function presenceSigningBytes(inp) {
  const p = [];
  field(p, inp.protocol);
  field(p, inp.alg);
  field(p, inp.id);
  field(p, hexBuf(inp.public_key_hex));
  field(p, inp.endpoint);
  field(p, inp.mcp_path);
  const caps = [...inp.capabilities].sort();
  u32(p, caps.length);
  for (const c of caps) field(p, c);
  u32(p, inp.protocol_major);
  field(p, inp.grant_id);
  u64(p, inp.heartbeat_unix);
  return Buffer.concat(p);
}

function grantSigningBytes(inp) {
  const p = [];
  field(p, "J3nna-mesh-grant/1");
  field(p, inp.alg);
  field(p, inp.id);
  field(p, inp.subject);
  field(p, hexBuf(inp.public_key_hex));
  u64(p, inp.tier);
  field(p, [...inp.scopes].sort().join("\0"));
  u64(p, inp.issued_at);
  u64(p, inp.not_after);
  if (inp.principal) {
    field(p, "J3nna-mesh-principal/1");
    field(p, inp.principal);
  }
  return Buffer.concat(p);
}

// Reproduce Go's json.Marshal of a map: keys sorted, compact, and <, >, & escaped as < > &.
function canonicalArgsJSON(args) {
  const sortKeys = (v) =>
    v && typeof v === "object" && !Array.isArray(v)
      ? Object.fromEntries(Object.keys(v).sort().map((k) => [k, sortKeys(v[k])]))
      : Array.isArray(v)
      ? v.map(sortKeys)
      : v;
  return JSON.stringify(sortKeys(args))
    .replace(/</g, "\\u003c")
    .replace(/>/g, "\\u003e")
    .replace(/&/g, "\\u0026");
}

function callproofSigningBytes(inp) {
  const p = [];
  field(p, "JIP-call/0.2");
  field(p, inp.alg);
  field(p, inp.node_id);
  field(p, inp.tool);
  field(p, hexBuf(inp.args_hash_hex));
  u64(p, inp.unix_milli);
  return Buffer.concat(p);
}

const builders = {
  "presence-record": presenceSigningBytes,
  grant: grantSigningBytes,
  callproof: callproofSigningBytes,
};

function ed25519PublicKey(hex) {
  // import a raw 32-byte ed25519 public key via JWK (base64url, no padding)
  const x = Buffer.from(hex, "hex").toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
  return crypto.createPublicKey({ key: { kty: "OKP", crv: "Ed25519", x }, format: "jwk" });
}

function main() {
  if (vectors.protocol !== "JIP/0.1") throw new Error(`unexpected protocol ${vectors.protocol}`);
  for (const v of vectors.vectors) {
    const build = builders[v.name];
    if (!build) throw new Error(`no Node builder for vector ${v.name}`);

    const got = build(v.input);
    if (got.toString("hex") !== v.signing_bytes_hex) throw new Error(`${v.name}: signing bytes differ`);

    const pub = ed25519PublicKey(v.signer_public_key_hex);
    const ok = crypto.verify(null, got, pub, Buffer.from(v.signature_b64, "base64"));
    if (!ok) throw new Error(`${v.name}: signature did not verify`);

    if (v.name === "callproof") {
      const cj = canonicalArgsJSON(v.input.args);
      if (cj !== v.input.args_canonical_json) throw new Error("args canonical JSON differs from Go");
      const h = crypto.createHash("sha256").update(cj).digest("hex");
      if (h !== v.input.args_hash_hex) throw new Error("args hash differs");
    }
    console.log(`  ok  ${v.name}`);
  }
  console.log(`PASS: ${vectors.vectors.length} vectors verified (Node.js wire-compatible with the Go reference)`);
}

main();
